### PR TITLE
[CRR] Support Endpoint Discovery when available

### PR DESCRIPTION
**Pull request by cjyclaire**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

will do another final review of the diff before requesting review :) pending changelog entry

### CODE REVIEW

**Comment by cjyclaire** [comment]
will follow up with unit test fixes

**Comment by cjyclaire** [comment]
Tests fixed, changelog added, ready for final review.

File: build_tools/aws-sdk-code-generator/lib/aws-sdk-code-generator/views/client_api_module.rb [hunk 1]
```diff
@@ -450,6 +468,15 @@ def initialize
         # @return [Boolean]
         attr_accessor :deprecated
 
+        # @return [Boolean]
+        attr_accessor :endpoint_discovery_exist
```
Comment (awood45) [inline build_tools/aws-sdk-code-generator/lib/aws-sdk-code-generator/views/client_api_module.rb]: This isn't very descriptive...

`:endpoint_discovery_required`?
`:endpoint_discovery_available`?

File: gems/aws-sdk-core/CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,8 @@
 Unreleased Changes
 ------------------
 
+* Feature - Adding support for endpoint discovery for service operations when available, to enable this feature, set `:endpoint_discovery` to `true` for the client.
```
Comment (awood45) [inline gems/aws-sdk-core/CHANGELOG.md]: Perhaps worth noting that not all clients support this?

**Comment by cjyclaire** [comment]
Integ tests added also added dynamo code-gen diff, feedbacks addressed, a final check? :) @awood45

**Pull request merged**